File: src/Aiko.Core/CredentialFile.cs
using System.Text.Json;

namespace Aiko.Core;

/// What Claude Code keeps in .credentials.json. Parsing only: the file is read by the app,
/// kept in memory for one request and never written anywhere by us.
public sealed class CredentialFile
{
    private CredentialFile(string accessToken, DateTimeOffset expiresAt, string subscriptionType)
    {
        AccessToken = accessToken;
        ExpiresAt = expiresAt;
        SubscriptionType = subscriptionType;
    }

    public string AccessToken { get; }
    public DateTimeOffset ExpiresAt { get; }
    public string SubscriptionType { get; }

    public bool IsExpiredAt(DateTimeOffset now) => ExpiresAt <= now;

    public static bool TryParse(string json, out CredentialFile? credential)
    {
        credential = null;
        if (string.IsNullOrWhiteSpace(json))
        {
            return false;
        }

        try
        {
            using var document = JsonDocument.Parse(json);
            if (document.RootElement.ValueKind != JsonValueKind.Object
                || !document.RootElement.TryGetProperty("claudeAiOauth", out var oauth)
                || oauth.ValueKind != JsonValueKind.Object)
            {
                return false;
            }

            if (!oauth.TryGetProperty("accessToken", out var token) || token.ValueKind != JsonValueKind.String
                || !oauth.TryGetProperty("expiresAt", out var expires) || expires.ValueKind != JsonValueKind.Number)
            {
                return false;
            }

            var accessToken = token.GetString();
            if (string.IsNullOrEmpty(accessToken))
            {
                return false;
            }

            var subscription = oauth.TryGetProperty("subscriptionType", out var type) && type.ValueKind == JsonValueKind.String
                ? type.GetString() ?? "unknown"
                : "unknown";

            credential = new CredentialFile(
                accessToken,
                DateTimeOffset.FromUnixTimeMilliseconds(expires.GetInt64()),
                subscription);
            return true;
        }
        catch (JsonException)
        {
            return false;
        }
    }

    /// Never print the token. Logs, exceptions and diagnostics all end up calling this.
    public override string ToString() => $"Credential({SubscriptionType}, expires {ExpiresAt:u}, token hidden)";
}

File: mac/Sources/AikoKit/CredentialFile.swift
import Foundation

/// What Claude Code keeps in .credentials.json. Parsing only: the file is read by the app,
/// kept in memory for one request and never written anywhere by us.
public struct CredentialFile: Sendable, Equatable, CustomStringConvertible {
    public let accessToken: String
    public let expiresAt: Date
    public let subscriptionType: String

    private init(accessToken: String, expiresAt: Date, subscriptionType: String) {
        self.accessToken = accessToken
        self.expiresAt = expiresAt
        self.subscriptionType = subscriptionType
    }

    public func isExpiredAt(_ now: Date) -> Bool { expiresAt <= now }

    /// Nil when this is not a credential file. `TryParse` with an out parameter on Windows.
    public static func parse(_ json: String?) -> CredentialFile? {
        guard let json, !json.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              let root = JsonNode.parse(json)?.objectValue,
              let oauth = root["claudeAiOauth"]?.objectValue else {
            return nil
        }

        guard let token = oauth["accessToken"]?.stringValue, !token.isEmpty,
              let expires = oauth["expiresAt"]?.int64Value else {
            return nil
        }

        return CredentialFile(
            accessToken: token,
            expiresAt: Date(timeIntervalSince1970: Double(expires) / 1000),
            subscriptionType: oauth["subscriptionType"]?.stringValue ?? "unknown")
    }

    /// Never print the token. Logs, errors and diagnostics all end up calling this.
    public var description: String {
        "Credential(\(subscriptionType), expires \(Iso8601.roundTrip(expiresAt, offsetSeconds: 0)), token hidden)"
    }
}
